### PR TITLE
Write optimal moran player experiments

### PR DESCRIPTION
**Pull request by drvinceknight**
There are essentially two pieces of work here.

1. I had to modify the environment slightly because of some upstream dependency changes see: 016f3d0. I also modified `get_evolutionary_best_response` to be able to take a `K` number of self interactions. As part of that I added a number of tests, could you take a close look please @Nikoleta-v3 as I'm not sure the behaviour make sense (see: fb3835d and 
7ac441f)

2. I implemented the script to carry out the Moran Process. There are a couple of questions for us to figure out:
    1. Is this working correctly? 
    2. Is it in the right place (I think `data/` is in the gitigore) and are we happy with `csv` output (the rest of the output data is sql).

I'll get this running now on siren but as you review we can rerun.

FWIW here is a output of the data so far:

```csv
seed,Opponent_p1,Opponent_p2,Opponent_p3,Opponent_p4,K,Best_response_p1,Best_response_p2,Best_response_p3,Best_response_p4,A_11,A_12,A_21,A_22,x_K,non_dynamic_x_K
0,0.5488135039273248,0.7151893663724195,0.6027633760716439,0.5448831829968969,1,1.0,3.6269121069924184,0.3432719732518952,2.4396306669294705,0.05783263599610761,0.19097371258391127,0.0,0.0,0.8179014384844345,0.5899545930905026
0,0.5488135039273248,0.7151893663724195,0.6027633760716439,0.5448831829968969,2,1.0,3.6269121069924184,0.34327197325189523,2.4396306669294705,0.01407327223347866,0.13121794942375484,0.0,0.0,0.9675906296654478,0.8731208624686253
0,0.5488135039273248,0.7151893663724195,0.6027633760716439,0.5448831829968969,3,1.0,3.6269121069924184,0.3432719732518953,2.4396306669294705,0.0,0.14238760031269163,0.0,0.0,0.9949861725782292,0.9803714144261786
1,0.417022004702574,0.7203244934421581,0.00011437481734488664,0.30233257263183977,1,1.0,3.077858252777325,0.4805354368056689,1.9861929749629406,0.0,0.24939570935833372,0.0,0.0,0.7207819344245042,0.5756150473040469
1,0.417022004702574,0.7203244934421581,0.00011437481734488664,0.30233257263183977,2,1.0,3.077858252777325,0.48053543680566857,1.9861929749629406,0.0,0.22678639593704739,0.0,0.0,0.9254122595291124,0.8532072715237938
1,0.417022004702574,0.7203244934421581,0.00011437481734488664,0.30233257263183977,3,1.0,3.0778582527773253,0.4805354368056687,1.9861929749629406,0.0,0.20550240743784606,0.0,0.0,0.9835069036267767,0.9675406896312749
2,0.43599490214200376,0.025926231827891333,0.5496624778787091,0.4353223926182769,1,1.0,2.2354862450431234,0.6911284387392193,1.9774534720198647,0.10159318306299141,0.07346496961142561,0.0,0.0,0.5415731620092256,0.4298380603410497
2,0.43599490214200376,0.025926231827891333,0.5496624778787091,0.4353223926182769,2,1.0,2.235486245043123,0.691128438739219,1.9774534720198647,0.0,0.18934955926122052,0.0,0.0,0.8066879036948199,0.7276170993365183
2,0.43599490214200376,0.025926231827891333,0.5496624778787091,0.4353223926182769,3,1.0,2.2354862450431234,0.6911284387392193,1.9774534720198647,0.0,0.5780253145871854,0.0,0.0,0.9364687554605542,0.9104824529806129
3,0.5507979025745755,0.7081478226181048,0.2909047389129443,0.510827605197663,1,1.0,3.545610921236985,0.3635972696907537,2.3017146502139116,0.0,0.22652114455914577,0.0,0.0,0.804509591187406,0.5966038249828565
3,0.5507979025745755,0.7081478226181048,0.2909047389129443,0.510827605197663,2,1.0,3.545610921236985,0.3635972696907538,2.3017146502139116,0.0,0.18228456985021482,0.0,0.0,0.9627522960671692,0.8751968156176341
3,0.5507979025745755,0.7081478226181048,0.2909047389129443,0.510827605197663,3,1.0,3.545610921236985,0.3635972696907538,2.3017146502139116,0.0,0.17852200243499644,0.0,0.0,0.9938777840936143,0.9794867344851333
4,0.9670298390136767,0.5472322491757223,0.9726843599648843,0.7148159936743647,1,2.6166211535388046,3.109906238648952,2.3868498865466607,2.9419253789544273,0.5928446182250184,0.8442657485810175,0.857945617622757,0.8472517387841256,nan,0.3007637913961168
4,0.9670298390136767,0.5472322491757223,0.9726843599648843,0.7148159936743647,2,1.0,3.4488727152717265,0.3877818211820842,2.9419253789544273,1.0,0.821724533197953,0.0,0.0,nan,0.8226613257281965
4,0.9670298390136767,0.5472322491757223,0.9726843599648843,0.7148159936743647,3,nan,3.448872715271749,0.3877818211820791,2.9419253789544273,1.0,0.6624061443294316,0.0,0.0,nan,nan
5,0.22199317108973948,0.8707323061773764,0.20671915533942642,0.9186109079379216,1,1.0,4.50655469591443,0.12336132602139249,2.30451618699129,0.0,0.34429805854655876,0.0,0.0,0.9431312318105877,0.692378516806652
5,0.22199317108973948,0.8707323061773764,0.20671915533942642,0.9186109079379216,2,1.0,4.50655469591443,0.12336132602139255,2.30451618699129,0.0,0.30392452360478167,0.0,0.0,0.996775253019569,0.9347371811705407
5,0.22199317108973948,0.8707323061773764,0.20671915533942642,0.9186109079379216,3,1.0,4.50655469591443,0.12336132602139256,2.30451618699129,0.0,0.2730700711420934,0.0,0.0,0.9998264519560575,0.9964877137280047
6,0.8928601514360016,0.3319798053011772,0.8212291230578318,0.0416966257252499,1,3.0,2.6481082932830646,3.139800489803071,2.051095741478738,1.0,0.6502648608072523,1.0,1.0,0.3030462078622568,0.26150652638466476
6,0.8928601514360016,0.3319798053011772,0.8212291230578318,0.0416966257252499,2,nan,2.3873024080718164,2.7361614161432732,2.051095741478738,1.0,0.623512495021336,1.0,0.0,0.5675669546482449,nan
6,0.8928601514360016,0.3319798053011772,0.8212291230578318,0.0416966257252499,3,1.0,1.2350042976314703,0.9412489255921324,2.051095741478738,0.0,0.490300500055603,0.0,0.0,0.7984598800994711,0.7435038195288733
```

### CODE REVIEW

File: moran_data/moran/main.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    )
+
+
+def obtain_payoff_matrix(opponent, K, N=4):
+    """
+    Obtain the payoff matrix for K best response players in a population of a
+    total of N individuals.
+    """
+    best_ev_response, _, _ = opt_mo.get_evolutionary_best_response(
+        [opponent], opt_mo.get_memory_one_best_response, K=K
```
Comment (drvinceknight) [inline moran_data/moran/main.py:38]: This needs to be `[opponent] * (N - K)`

```suggestion
        [opponent] * (N - K), opt_mo.get_memory_one_best_response, K=K
```

File: moran_data/moran/main.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        payoff_matrices[K] = payoff_matrix
+        best_response_players[K] = best_response
+
+    f_ones = np.array(
+        [
+            (
+                payoff_matrices[K][0, 0] * (K - 1)
+                + payoff_matrices[K][0, 1] * (N - K)
+            )
+            for i in range(1, N)
```
Comment (drvinceknight) [inline moran_data/moran/main.py:68]:
```suggestion
            for K in range(1, N)
```

File: moran_data/moran/main.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        ]
+    )
+    f_twos = np.array(
+        [
+            (
+                payoff_matrices[K][1, 0] * K
+                + payoff_matrices[K][1, 1] * (N - K - 1)
+            )
+            for i in range(1, N)
```
Comment (drvinceknight) [inline moran_data/moran/main.py:77]:
```suggestion
            for K in range(1, N)
```

File: moran_data/moran/main.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        with open("data.csv", "a") as f:
+            csv_writer = csv.writer(f)
+
+            np.random.seed(seed)
+            opponent = np.random.random(4)
+            (
+                fixation_probabilities,
+                payoff_matrices,
+                best_response_players,
+            ) = best_response_moran_process(opponent, N=4)
```
Comment (drvinceknight) [inline moran_data/moran/main.py:134]:
```suggestion
            ) = best_response_moran_process(opponent, N=N)
```

File: moran_data/moran/main.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+            row = [seed] + list(opponent)
+            for K in range(1, 4):
```
Comment (drvinceknight) [inline moran_data/moran/main.py:137]:
```suggestion
            for K in range(1, N):
```

File: moran_data/moran/README.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+Seed, Opponent p1, Opponent p2, Opponent p3, Opponent p4, K, Best response p1, Best response p2, Best response p3, Best respose p4, A_11, A_12, A_21, A_22, x_k, non_dynamic_x_k
+```
+
+Approach:
+
+The fixation probability \\(x_i\\) is given by:
+
+\\[
+    x_k = \frac{
+            1 + \sum_{j=1}^{j-1}\prod_{i=1}^j\gamma_i
```
Comment (Nikoleta-v3) [inline moran_data/moran/README.md:25]:
```suggestion
            1 + \sum_{j=1}^{k-1}\prod_{i=1}^j\gamma_i
```
it should be this no?

File: moran_data/moran/main.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        ]
+    )
+    f_twos = np.array(
+        [
+            (payoff_matrix[1, 0] * K + payoff_matrix[1, 1] * (N - K - 1))
+            for K in range(1, N)
+        ]
+    )
+    gammas = f_twos / f_ones
+    return (1 + np.sum(np.cumprod(gammas[: K - 1]))) / (
```
Comment (Nikoleta-v3) [inline moran_data/moran/main.py:27]: `K` at this point should be 1, thus `gammas[: K - 1]` is just `gammas[:0]`. No?

File: moran_data/moran/README.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+The data set is of the form:
+
+```
+Seed, Opponent p1, Opponent p2, Opponent p3, Opponent p4, K, Best response p1, Best response p2, Best response p3, Best respose p4, A_11, A_12, A_21, A_22, x_k, non_dynamic_x_k
+```
+
+Approach:
+
+The fixation probability \\(x_i\\) is given by:
```
Comment (Nikoleta-v3) [inline moran_data/moran/README.md:21]: here we state that \(x_i\) is given by then use  \(x_k\) notation

**Review by Nikoleta-v3 — COMMENTED** [review]
I have gone through the `moran` code now carefully @drvinceknight 👍  everything looks right to me and you are calculating the fixation probabilities correctly. 

I just have a minor comment on the readme, not super important right now.

Will have a closer looks at the tests and then we should be good to merge.

**Comment by Nikoleta-v3** [comment]
@drvinceknight once the tests pass this should be ready for review 👍

File: tests/test_evolutionary_best_response.py [hunk 9]
```diff
@@ -232,9 +232,7 @@ def test_get_evolutionary_best_response_with_K_equal_10_against_2_defectors():
     With K = 10 and 2 defectors the best response changes from `[1, 0.45420466,
     0, 0.03799109]` (for K = 1) to:
 
-    [0.26874769, 0.11697721, 0, 0]
-
-    TODO: Nik does this make sense?
+    [1, 0.36819048, 1, 0.75011483]
```
Comment (drvinceknight) [inline tests/test_evolutionary_best_response.py:235]: This is better right? This makes more sense?

File: src/opt_mo/evolutionary_best_response.py [hunk 10]
```diff
@@ -49,4 +50,11 @@ def get_evolutionary_best_response(
         history.append(best_response)
         cycle, repeat_length = get_repeat_cycle_and_length(history, tol=tol)
 
+    utilities = [
+        opt_mo.tournament_utility(player, opponents)
+        + opt_mo.tournament_utility(player, [player] * K)
```
Comment (drvinceknight) [inline src/opt_mo/evolutionary_best_response.py:55]: Is the following slightly nicer?

```suggestion
        opt_mo.tournament_utility(player, opponents + [player] * K)
```

**Comment by drvinceknight** [comment]
I've fixed those things in the README. Let me know what you think about my other suggestion: I can do that and push. Once that's done I'll get the data running again.

File: tests/test_evolutionary_best_response.py (hunk 9 above)
Comment (Nikoleta-v3) [reply tests/test_evolutionary_best_response.py:235]: Yes it does!

Comment (drvinceknight) [reply tests/test_evolutionary_best_response.py:235]: 🎉

**Pull request merged**